src/services/datamigration/datamigrationService.js

```diff
@@ -0,0 +1,194 @@
+						});
+					}));
+				});
+
+				Promise.all(valuesPromises).then((data) => {
+					this.saveAllData(topics, data);
+				});
+			})
+		});
+	}
+
+	/**
+	 * saveAllData
+	 * 
+	 * Saves data from the old storage adapter into the new one.
+	 * 
+	 * Invoked when all data has been retrieved from the old adapter.
+	 * @param {Array} topics Array of all topics retrieved.
+	 * @param {Array} data Array of all data received from retrieveDatabase
+	 */
+	saveAllData(topics, dataset) {
+		const promiseSet = [];
+
+		topics.forEach((topic) => {
+			promiseSet.push(new Promise((resolve, reject) => {
+				StorageClient.setStore({ topic: topic, dataStore: this.newAdapter }, (err, data) => {
+					if (err) {
+						reject(new Error(err));
+					}
+					resolve(data);
+				});
+			}));
+		});
+
+		Promise.all(promiseSet).then(() => {
+			const saveSet = [];
+			
+			saveSet.push(new Promise((resolve, reject) => {
+
+				dataset.forEach((datum) => {
+					StorageClient.save({ topic: datum.topic, key: datum.key, value: datum.value }, (err, response) => {
+						if (err) {
+							reject(new Error(err));
+						}
+						resolve(response);
+					});
+				});
+			}));
+
+			Promise.all(saveSet).then(() => {
+				StorageClient.save({ topic: "finsemble", key: `migrated_from_${this.adapter}`, value: Date.now() });
+				RouterClient.transmit("Migration", "end");
+			});
+		});
+	}
+
+	/**
+	 * fetchUserStatus
+	 * 
+	 * Determines if the user has been migrated from IndexedDBAdapter (or another storage adapter)
+	 * by checking for a key in the new storage service of "migrated_from_<adapter to migrate from>", 
+	 * which in this example is "migrated_from_IndexedDBAdapter".
+	 */
+	fetchUserStatus() {
+		StorageClient.get({ topic: "finsemble", key: `migrated_from_${this.adapter}` }, (err, data) => {
+			if (err) {
+				throw new Error(err);
+			}
+
+			if (!data) {
+				// The user does not have a migration record.
+				
+				StorageClient.get({ topic: "finsemble.workspace.cache", key: "activeWorkspace"}, (err, data) => {
+					if (err) {
+						throw new Error(err);
+					}
+
+					if (data) {
+						// The user has a default workspace and thus should be migrated.
+
+						LauncherClient.spawn("migration", {}, (err, response) => {
+							RouterClient.transmit('Migration', 'needed');
+							RouterClient.addListener('Migration', (err, event) => {
+								if (event.data == 'begin') {
+									Logger.log("*** datamigration begin")
+									this.retrieveDatabase(TOPICS);
+								}
+							});
+						});
+					} else {
+						// This is a new user. Set their migration record as 0 so they're not checked again.
+						StorageClient.save({ topic: "finsemble", key: `migrated_from_${this.adapter}`, value: 0 });
+						RouterClient.transmit("Migration", "not needed");
+					}
+				});
+			} else {
+				RouterClient.transmit("Migration", "not needed");
+			}
+		});
+	}
+}
+
+const TOPICS = ['finsemble', 'finsemble.workspace', 'finsemble.workspace.cache'];
```

Comment:
It might not make sense to transfer `finsemble.workspace.cache`. That should stay as local (either IndexedDB or localstorage) because it is hit _very_ frequently.